feat(Email Trigger (IMAP) Node): IMAP trigger node returns message UIDs

umanamente opened the pull request:
## Summary

This PR adds returning the UID for each email message in the IMAP Trigger node. This allows subsequent nodes to use the UID for email manipulation. For example, the community IMAP node [n8n-nodes-imap](https://github.com/umanamente/n8n-nodes-imap) supports moving and copying emails between mailboxes, which requires the UID. Previously, this information was missing in the IMAP Trigger node output.

### Steps to test:
1. Configure the IMAP Trigger node.
2. Wait for it to return a new email message.
3. Check the output data — the `attributes` field should contain the `UID` field.

![2025-02-09 12_49_53](https://github.com/user-attachments/assets/f3b71e83-2066-4707-800a-8ff677dd68a7)


## Related Linear tickets, Github issues, and Community forum posts

- Forum feature request [IMAP Trigger Node - Provide email UIDs](https://community.n8n.io/t/imap-trigger-node-provide-email-uids/75626)
- GitHub issue: https://github.com/umanamente/n8n-nodes-imap/issues/42

## Review / Merge checklist

- [x] PR title and summary are descriptive.  
- [x] Docs updated or follow-up ticket created.  NO DOCS AVAILABLE IN BASELINE
- [x] Tests included.  - NO TESTS AVAILABLE IN BASELINE
- [x] PR labeled with `release/backport` (if necessary). - NO NEED

umanamente committed:
feat(Email Trigger (IMAP) Node): IMAP trigger node returns message UIDs

Joffcom commented:
Hey @umanamente,

Thanks for the PR, We have created "GHC-783" as the internal reference to get this reviewed.

One of us will be in touch if there are any changes needed, in most cases this is normally within a couple of weeks but it depends on the current workload of the team.

dana-gill commented:
@umanamente hi, could you please fix the tests 🙏 thank you!

dana-gill commented on packages/nodes-base/nodes/EmailReadImap/test/v2/utils.test.ts:
```diff
@@ -42,6 +42,9 @@ describe('Test IMap V2 utils', () => {
 							headerLines: undefined,
 							html: false,
+							attributes: {
+								uid: undefined,
```
Could you please add the following test cases:
1. Expect the `uid` to be defined
2. `uid` is in the wrong format -- Probably requires some `try/catch` in `getEmails` of `v2`.

🙏

umanamente replied:
I added the first test: 'Expect the uid to be defined.'
I think the second test is unnecessary since the uid is already used in the code. Let me know your thoughts.

dana-gill commented on packages/nodes-base/nodes/EmailReadImap/v1/EmailReadImapV1.node.ts:
```diff
@@ -435,6 +435,9 @@ export class EmailReadImapV1 implements INodeType {
 							textHtml: await getText(parts, message, 'html'),
 							textPlain: await getText(parts, message, 'plain'),
 							metadata: {} as IDataObject,
+							attributes: {
+								uid: message.attributes.uid,
+							} as IDataObject,
```
```suggestion
```

We would like to avoid adding new features for older versions of the node

dana-gill commented:
Hey @umanamente thanks for adding the test case for `uid` to be defined 😄 I think there was a bit of miscommunication on my end for what I would expect to see for test cases. Here are the test cases I would expect to see for this PR:

1. `uid` is not defined
2. `uid` is defined

So you would have the following tests cases:

```
it('should return new emails' => ... )
// keep as is

it('should return new emails with uid if uid is present' => ... )
// this should be the same 'should return new emails' except you add uid to message and expect uid
```

Let me know if this makes sense

umanamente commented:
If uid is not present, the node itself will break, since uid is used in the code before returning

koordinator7 commented:
\ @umanamente 

Good day. Please tell me how I can get UID in the IMAP node? If it's not too much trouble, in more detail, where and what code should I write?

umanamente commented:
@dana-gill 
I don't think adding a test to validate the UID makes sense because if the UID is not valid, the existing node functionality will break. If this test is necessary, it should be added by the developers maintaining this node, as it affects the node's functionality itself rather than the changes introduced in this pull request.  

Please let me know your thoughts and if any further changes are needed before it can be merged.

koordinator7 commented:
@umanamente 
 Good day. Please tell me how I can get UID in the IMAP node? If it's not too much trouble, in more detail, where and what code should I write?

umanamente commented:
> @umanamente 
>  Good day. Please tell me how I can get UID in the IMAP node? If it's not too much trouble, in more detail, where and what code should I write?
> 
> 

This function is not available in the node yet. This pull request adds it, but it needs to be merged first.

koordinator7 commented:
@umanamente
> This function is not available in the node yet. This pull request adds it, but it needs to be merged first.
Thank you. I understand that the standard IMAP node does not yet contain UID. But you somehow get UID in the IMAP node. Please tell me what I need to do to get UID too. If it is not difficult, please provide more details.

dana-gill commented:
Hey @umanamente could you rebase off of master and then ping me and you've done that? Thanks

umanamente commented:
> Hey @umanamente could you rebase off of master and then ping me and you've done that? Thanks

@dana-gill , sure, done

dana-gill commented:
@umanamente it looks like the unit tests are failing. could you please patch them up and then ping me again when the pr is ready? thanks!

koordinator7 commented:
Hey @umanamente
Help!!! "This function is not available in the node yet". This pull request adds it, but it needs to be merged first.
Thank you. I understand that the standard IMAP node does not yet contain UID. But you somehow get UID in the IMAP node. Please tell me what I need to do to get UID too. If it is not difficult, please provide more details.

umanamente committed:
fixed unit tests: reset staticData before each execution

umanamente commented:
@dana-gill ,
I’ve fixed the unit tests. The issue was that the same `staticData` object was being reused across different runs. During the first run, the function added the last email's ID to `staticData`, and since that ID persisted between runs, the email was getting filtered out, resulting in an empty array. I’ve updated it so that a fresh `staticData` object is created before each run.

Could you please run the tests again?

dana-gill commented on packages/nodes-base/nodes/EmailReadImap/test/v2/utils.test.ts:
```diff
@@ -74,6 +73,9 @@ describe('Test IMap V2 utils', () => {
 			];
 
 			expectedResults.forEach(async (expectedResult) => {
+				// use new staticData for each iteration
+				const tmpStaticData: IDataObject = {};
```
```suggestion
				const staticData: IDataObject = {};
```

dana-gill commented on packages/nodes-base/nodes/EmailReadImap/test/v2/utils.test.ts:
```diff
@@ -85,7 +87,7 @@ describe('Test IMap V2 utils', () => {
 					imapConnection,
 					[],
-					staticData,
+					tmpStaticData,
```
```suggestion
					staticData,
```

umanamente committed:
renamed staticData

dana-gill commented:
> @dana-gill Could you please let me know what the next steps are and if anything is needed from my side for this Pull Request to be merged?

merged ;)

janober commented:
Got released with [`n8n@1.88.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.88.0)

Alexandero89 commented:
@umanamente thanks for your PR and the time you invested.

i just noticed that the UID is only available when i choose the format of the mails to be "Simple".
When i choose the format "Resolved" there is not attributes and no uid.
Was this planned and is this maybe a technical restriction or would it maybe be possible to extend it to the resolved format.

Thanks in advance

Alexandero89 commented:
Ok, i created a PR with this feature: https://github.com/n8n-io/n8n/pull/15539
